vars/pythonSetupPyPipeline.groovy

```diff
@@ -9,148 +9,310 @@
  */
 
 def call(Map pipelineParams) {
-  def LOG_TAG = "[pythonPipeline]"
 
-  if (!pipelineParams.pypiCredentials) {
-    error("${LOG_TAG} Please provide pipelineParams.pypiCredentials")
-  }
-  if (!pipelineParams.pypiRepo) {
-    error("${LOG_TAG} Please provide pipelineParams.pypiRepo")
-  }
-  if (!pipelineParams.sshAgentUser) {
-    error("${LOG_TAG} Please provide pipelineParams.sshAgentUser")
-  }
-  if (!pipelineParams.dockerFilename) {
-    pipelineParams["dockerFilename"] = "Dockerfile"
-  }
-  if (!pipelineParams.dockerBuildArgs) {
-    pipelineParams["dockerBuildArgs"] = ""
-  }
-  if (!pipelineParams.dockerRunArgs) {
-    pipelineParams["dockerRunArgs"] = "-v /etc/passwd:/etc/passwd:ro -v /opt/jenkins/.ssh:/opt/jenkins/.ssh:ro --network host"
-  }
+  validateParameter(pipelineParams)
+  initParameterWithBaseValues(pipelineParams)
+  log("Pipeline params: ${pipelineParams}")
 
   pipeline {
     agent {
-      dockerfile {
-        filename pipelineParams.dockerFilename
-        additionalBuildArgs pipelineParams.dockerBuildArgs
-        args pipelineParams.dockerRunArgs
+      node {
+        label "docker"
       }
     }
 
     parameters {
-      booleanParam(defaultValue: false, description: 'Release the module', name: 'doRelease')
+      booleanParam(defaultValue: false, description: "Release the module", name: "doRelease")
+      booleanParam(defaultValue: false, description: "Deploy snapshot Docker image", name: "doDockerSnapshot")
     }
 
     options {
       disableConcurrentBuilds()
+      buildDiscarder(logRotator(numToKeepStr: '5'))
     }
 
     stages {
+
+      // We create the build image only once and use it in later stages (reduces job time by ~50%).
+      stage("Create build image") {
+        steps {
+          script {
+            buildImage = docker.build("python-setup-py-build", "-f ${pipelineParams.dockerBuildFile} ${pipelineParams.dockerBuildArgs} .")
```

Comment:
Why not consider a multi-stage build docker file and then just stop here build at specific configured stage?

Reply:
Thanks a lot for this suggestion. Addressed your input and updated this PR.